enhancements/security/security-profiles-operator.md

```diff
@@ -0,0 +1,636 @@
+to do the actual recording. So we'd need to ensure this hook is available in RHCOS
+before enabling this feature.
+
+There isn't currently support for `SelinuxProfile` recording, but
+the work is scheduled to start in the 4.9 cycle, and the intent
+is to leverage [Udica](https://github.com/containers/udica) to generate
+the profile.
+
+### SCC integration
+
+Security Context Constraints allow a user to restrict the security settings
+that a workload may have. Given that the profile settings are in the pod's
+`securityContext` it is imperative that we appropriately integrate with SCCs.
+
+Currently, the only SCC that allows users to set the Seccomp and SELinux
+settings for a pod is the privileged one. It looks as follows:
+
+```bash
+Name:                                           privileged
+Priority:                                       <none>
+Access:                                         
+  Users:                                        system:admin,system:serviceaccount:openshift-infra:build-controller
+  Groups:                                       system:cluster-admins,system:nodes,system:masters
+Settings:                                       
+  Allow Privileged:                             true
+  Allow Privilege Escalation:                   true
+  Default Add Capabilities:                     <none>
+  Required Drop Capabilities:                   <none>
+  Allowed Capabilities:                         *
+  Allowed Seccomp Profiles:                     *
+  Allowed Volume Types:                         *
+  Allowed Flexvolumes:                          <all>
+  Allowed Unsafe Sysctls:                       *
+  Forbidden Sysctls:                            <none>
+  Allow Host Network:                           true
+  Allow Host Ports:                             true
+  Allow Host PID:                               true
+  Allow Host IPC:                               true
+  Read Only Root Filesystem:                    false
+  Run As User Strategy: RunAsAny                
+    UID:                                        <none>
+    UID Range Min:                              <none>
+    UID Range Max:                              <none>
+  SELinux Context Strategy: RunAsAny            
+    User:                                       <none>
+    Role:                                       <none>
+    Type:                                       <none>
+    Level:                                      <none>
+  FSGroup Strategy: RunAsAny                    
+    Ranges:                                     <none>
+  Supplemental Groups Strategy: RunAsAny        
+    Ranges:                                     <none>
+```
+
+This basicallly allows for setting all security settings for a pod, which
+might not be what a user intends to allow.
+
+To enable integration with the SPO and promote better security in workloads,
+we could add another SCC called `measured` that would look similar to this:
```

Comment:
@sttts Hi Stefan, I wonder if you have time to help us flesh out the design of the SCC part? Feel free to tag someone else if you're not the right person to ask of course..

What I'm thinking now is that it would be nice to have a more generic way for an SCC to allow a selinux context or a seccomp profile without having to use the privileged SCC and at the same time only restrict the security profile used to the same namespace as the workload -- as opposed to adding one more default SCC like the `measured` example below suggests. The one extra SCC wouldn't be flexible enough, I can see some workloads e.g. wanting to use only certain capabilities or only certain volume types etc. Moreover, there would still need to be a piece that checks that only security profiles from the same namespace are used and I wasn't sure if OCP would like to have a webhook that checks all pods cluster-wide (a validating webhook was my first though to restrict the workloads to a subset of security profiles).

Do you think there would be a way to allow only using a selinux or a seccomp profile based on values a project is annotated with, more or less like UID ranges are allocated now? Then the SCCs in general could be extended to allow SELinux context strategy "RunAsRange" that would look at some project annotation and a similar mechanism could be used to allow all seccomp profiles from a project. Also, this way, the SCC enforcement would be decoupled from the actual security profiles management, so this mechanism might work even if there's no SPO or its APIs.

What I don't know is, if it's possible to let some trusted component that maintains the security profiles (SPO in this case) to annotate the projects with the security profiles deployed there. I'm not super familiar with the project annotations, but a quick look shows that only the display name and description annotations can be modified.

What do you think?

Reply:
@deads2k I see you did a lot of work in the SCC are, so I wonder if you have any opinion on the above?

Reply:
@s-urbaniak @stlaz

Reply:
> The one extra SCC wouldn't be flexible enough, I can see some workloads e.g. wanting to use only certain capabilities or only certain volume types etc.

Isn't that why you can create your own? Or will this whole thing only ever work with the "measured" SCC?

> there would still need to be a piece that checks that only security profiles from the same namespace are used

Why limit security profiles to namespaces? Who gets to create them? Can I craft a malicious security profile in my namespace that would let me into other namespaces should any other service pick them up?

> would be a way to allow only using a selinux or a seccomp profile based on values a project is annotated with

This is already happening for SELinux (projects have the MCS part generated, although nothing else). You might be able to force seccomp on project's workloads in a similar fashion.

> I'm not super familiar with the project annotations, but a quick look shows that only the display name and description annotations can be modified.

I believe with enough privileges, you should be able to change any annotation on a namespace.

Reply:
> Can I craft a malicious security profile in my namespace that would let me into other namespaces should any other service pick them up?

That's right. It would be ideal if then you could only use security profiles that are "approved" in your namespace. This will help in both auditing, but also security by keeping a clear multi-tenancy.

Reply:
> > The one extra SCC wouldn't be flexible enough, I can see some workloads e.g. wanting to use only certain capabilities or only certain volume types etc.
> 
> Isn't that why you can create your own? Or will this whole thing only ever work with the "measured" SCC?

Yes, but the point of bringing the operator to OCP is that workloads would ship their own profiles. At the moment, the SCC allows you to list allowed profiles, or allow all or none. Creating an SCC per workload that ships a security profile doesn't seem very scalable. 

> 
> > there would still need to be a piece that checks that only security profiles from the same namespace are used
> 
> Why limit security profiles to namespaces? Who gets to create them? Can I craft a malicious security profile in my namespace that would let me into other namespaces should any other service pick them up?

Yes, that's the concern, especially with seccompProfiles. In the pod definition, `pods.spec.securityContext.seccompProfile` just points to a path on the node where the seccompProfile is stored. So the idea was that SPO would annotate projects with the seccomp or selinux profiles that are deployed there and SCC would enforce their use based on the annotations.

> 
> > would be a way to allow only using a selinux or a seccomp profile based on values a project is annotated with
> 
> This is already happening for SELinux (projects have the MCS part generated, although nothing else). You might be able to force seccomp on project's workloads in a similar fashion.
> 
> > I'm not super familiar with the project annotations, but a quick look shows that only the display name and description annotations can be modified.
> 
> I believe with enough privileges, you should be able to change any annotation on a namespace.

That's where I got confused, it [seemed](https://github.com/openshift/openshift-apiserver/blob/a6220d84588dd916fc13210d3484876832fd37bd/pkg/project/apis/project/validation/validation.go#L64) like right now only the description or displayname is allowed to be modified.

Of course, maybe this whole approach is over-engineered, because we'd have to add a new "strategy" for both selinux and seccomp that would allow all profiles from a list defined in a project, which is a big change. That's why I'm bringing this up even in a vague state of design.